### ISSUE TITLE
Unexpected behaviour in redirect next/navigation function - Next.js 15

### ISSUE THREAD

**Issue by leox182** [vercel/next.js#74258]
### Link to the code that reproduces this issue

https://github.com/leox182/next-unexpected-redirect-behavior

### To Reproduce

the repository has two projects, next-14 and next-15 that contain the same code with only difference between versions (14.2.21 and 15.1.2 respectively).
1. Install dependencies and start both projects in development mode.
2. Click on the ‘Say Hello Action!’ button in each and check the differences in behaviour when redirecting.

### Current vs. Expected behavior

Expected Behavior
Following the steps to reproduce, I expect the server action to execute its logic and then redirect the user without throwing an error (like on 14 version).

Actual Behavior
Instead of only redirecting, the server action throws an error, causing unexpected client-side behavior such as error messages.

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro for Workstations
  Available memory (MB): 32671
  Available CPU cores: 12
Binaries:
  Node: 22.12.0
  npm: 10.5.0
  Yarn: N/A
  pnpm: 9.15.1
Relevant Packages:
  next: 15.1.2 // Latest available version is detected (15.1.2).
  eslint-config-next: 15.1.2
  react: 19.0.0
  react-dom: 19.0.0
  typescript: 5.7.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Navigation

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

After upgrading to Next.js 15, I noticed a change in the behavior of the `redirect` function from `next/navigation` when used in a server action. 

In Next.js 14, calling `redirect` within a server action executes the action and then performs the redirect as expected. However, in Next.js 15, calling `redirect` not only redirects, but also causes the server action to throw an error that is cached on the client. This causes unexpected behaviour on the client, such as displaying an error message, even though no error has occurred.

While this can be handled on the client by checking if `error.message === "NEXT_REDIRECT"` and ignoring such errors, this change in behavior was not documented as a breaking change or improvement on [Next.js release post](https://nextjs.org/blog/next-15). Hence, I believe it is a bug.

**Comment by JoshuaKirby88** [comment]
NextJS throws a 303 error when you call redirect.
That's why when you call redirect in a try block inside a server action, it will be caught as an error, and will prevent you from redirecting.
The issue arises from the try block on the client catching the 303 error from the redirect.
But this wasn't a behaviour in Next14. In Next14, the server action will return undefined when you redirect, which can cause issues, because the return type for the server action will not reflect that.

I think in general this is an improvement because whatever code you have after the server action will not run when you redirect, which is what the return type suggests.

You can call "isRedirectError" from "next/dist/client/components/redirect-error":
```
    const handleSayHello = async () => {
        try {
            const res = await helloAction();
            console.log("res", res);

            alert("Said hello from the server!");
        } catch (error) {
            if (isRedirectError(error)) {
                throw error;
            }
            alert("Failed to say hello from the server!");
        }
    };
```

This way, "alert("Failed to say hello from the server!")" will only run when you encounter an actual error.
However, "alert("Said hello from the server!")" will not run anymore, unless you place it inside "if (isRedirectError(error)) {}" this block.

Either way, this is poor documentation by Vercel.

**Comment by Ikkevoid** [comment]
For me this problem only occurs on the Firefox browser, on Chromium based ones there's no problem like that.

**Comment by AnouarMc** [comment]
As stated in the [Next.js documentation](https://nextjs.org/docs/app/api-reference/functions/redirect), redirect() should be called outside of a try/catch block.

**Comment by leox182** [comment]
@JoshuaKirby88, Thanks for taking the time to answer, I have gone ahead with updating a project following a similar approach to yours, but the big problem is that this change in the redirect function behaviour was not documented, at least not clearly as far as I know.

> ```
>     const handleSayHello = async () => {
>         try {
>             await helloAction()
>         } catch (error) {
>             if (isRedirectError(error)) {
>                 alert("Said hello from the server!");
>             }
>             alert("Failed to say hello from the server!");
>         }
>     };
> ```

**Comment by leox182** [comment]
> As stated in the [Next.js documentation](https://nextjs.org/docs/app/api-reference/functions/redirect), redirect() should be called after the try/catch block.

@AnouarMc

the redirect function in the server action has been called after the try catch

**Comment by JoshuaKirby88** [comment]
> but the big problem is that this change in the redirect function behaviour was not documented, at least not clearly as far as I know.

I agree. 
This has been a glaring issue with Vercel for a while.

**Comment by AnouarMc** [comment]
@leox182 

Yes, but it still gets caught in the try/catch block of the page.js

```typescript
try {
  await helloAction();
  alert("Said hello from the server!");
} catch (error) {
  alert("Failed to say hello from the server!");
}
